packages/nimble-components/src/table/models/keyboard-navigation-manager.ts

```diff
@@ -0,0 +1,1025 @@
+/* eslint-disable no-console */
```

Comment:
Here are the set of behaviors that I noticed that I believe are not ideal (unexpected):

Table with hierarchy data (single selection mode):
- If the first action menu has focus, I think pressing Shift-Tab should focus the row.
- If an action menu has focus (or any focusable element), pressing the Left or Right arrow keys should shift focus to the previous or next cells respectively if there are no other focusable elements in the cell.
- If the table has ever moved focus to something like an action menu, then after you click out of the table, and then tab to it, it will skip the header content and refocus the action menu. This means that it is no longer easy to just tab past the table at the top-level.

Table Anchor Column
- The anchor column cells do not seem like they can show the highlighted cell border. I admit that this may be intentional, and not sure if there is a better answer.

Possibly more to come...

Reply:
Anchor Column: Intentional. This is the case of a cell containing a single focusable control, in which case we only focus the control, not the cell (according to the HLD).

I think our current anchor focus styling makes it a bit difficult to tell when the anchor is focused (at least compared to the focused cell styling), so we might end up revisiting that behavior. (I can demo it to the team and get opinions)

Reply:
I'm not sure if I agree on the 1st 3 points (behavior after an action menu is focused). To summarize the current behavior:  
Once an action menu is focused, you're no longer in row/cell navigation mode. You can return to that mode by pressing Esc, though. I'd characterize being able to press LeftArrow / RightArrow to go from a focused action menu to cell focus again as a behavior change from what's in the HLD, but if others agree I can look at adding that.

> If the table has ever moved focus to something like an action menu, then after you click out of the table, and then tab to it, it will skip the header content and refocus the action menu. This means that it is no longer easy to just tab past the table at the top-level.

I'm not sure how I'd reconcile that suggestion with our desire to return to the previously focused row once you tab past the table, and then tab back to it. Currently I'm characterizing interacting with the action menu the same way as interacting with a row/cell (in that we want to return focus to that same place when tabbing past and returning). I'm curious if others have an opinion on this point.

Reply:
>our desire to return to the previously focused row once you tab past the table, and then tab back to it

I think that workflow is important. It'd be annoying to navigate through a bunch of rows and columns to get to the cell you want, then accidentally press tab or click one extra time causing the table to lose focus and all your navigation to be lost.

I'm ok with the current behavior, but would be open to ideas for returning the table to its initial state where it only has one tab stop. Right now the way I know to do it is to reverse all the navigation that got you to your current state (ESC to unfocus action menu, arrows to return focus to row / header). As an example idea, maybe if you keep pressing ESC the focus should move up the hierarchy (action menu -> cell -> row -> header)?

But I'm wary of designing this on the fly and increasing scope too much before we get this in the hands of actual users.

Reply:
Yeah, I'm okay with the tab behavior. More than anything I wanted to check that we were fine with the `Table` entering into a state where it is no longer to just tab past it at the top-level. The AzDO table avoids this by not letting you tab through focusable elements in a row, and instead relies on the arrow keys (as it does not let you put cell focus in columns with no focusable elements). Screen readers like NVDA allow a user to navigate to non-focusable columns with Ctl-Arrow keys, which does not work without an active screen reader.

[This example](https://warsus.github.io/fancytree/demo/sample-aria-treegrid.html), (which I recognize is just some jo-shmo's idea of a treegrid) does what is currently implemented in ours.

For the first point on `Shift-Tab` putting focus on the row when the first action menu is focused, I think I just needed to become accustomed to pressing `Esc` to re-enter navigation mode.